plugins/inputs/ctrlx_datalayer_sse/ctrlx_datalayer_sse.go

```diff
@@ -0,0 +1,498 @@
+
+	requestBody := bytes.NewBuffer(p)
+	req, err := http.NewRequest("POST", sseURL, requestBody)
+	if err != nil {
+		s.Log.Errorf("Failed to create post request to create sse subscription %v, error: %v", sub.index, err)
+		return ""
+	}
+
+	req.Header.Add("Authorization", s.token.String())
+
+	resp, err := s.connection.Do(req)
+	if err != nil {
+		s.Log.Errorf("Failed to do post request to create sse subscription %v, error: %v", sub.index, err)
+		return ""
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != 200 && resp.StatusCode != 201 {
+		s.Log.Errorf("Failed to create sse subscription %v, status: %v", sub.index, resp.Status)
+		return ""
+	}
+
+	return sseURL + "/" + pl.Properties.ID
+}
+
+// reconnect is called when the connection is broken
+// reconnect starts restablishing the connection
+func (s *ctrlXDataLayerSSE) reconnect(sub subscription) {
+	go func() {
+		s.Log.Infof("Subscription %v is down, reconnect is starting", sub.index)
+		for {
+			time.Sleep(time.Duration(defaultReconnectInterval))
+
+			if !s.setAuthToken() {
+				continue
+			}
+
+			subURL := s.createSubscription(&sub)
+			if subURL == "" {
+				continue
+			}
+
+			client := newClient(subURL, sub, s.Log)
+			if s.autoReconnect {
+				client.onDisconnect(s.reconnect)
+			}
+			client.Headers["Authorization"] = s.token.String()
+			client.Connection.Transport = httpTransport(s.InsecureSkipVerify)
+			err := client.requestData(s, sub.Measurement)
+			if err != nil {
+				continue
+			}
+
+			s.Log.Infof("Subscription %v is reconnected", sub.index)
+			return
+		}
+	}()
+}
+
+// addMetric writes sse metric into accumulator
+func (s *ctrlXDataLayerSSE) addMetric(se *sSEEvent, measurement string) {
+	m := metric{}
+	switch se.Event {
+	case "update":
+		// Received an updated value, that we translate into a metric
+		_ = json.Unmarshal([]byte(se.Data), &m)
+		t := convertTimestamp2UnixTime(m.Timestamp)
+		if s.acc != nil {
+			s.wg.Add(1)
+			s.acc.AddFields(measurement, map[string]interface{}{"value": m.Value}, map[string]string{"node": m.Node, "type": m.Type, "host": s.Hostname}, t)
+			s.wg.Done()
+		}
+
+	case "error":
+		// Received an error event, that we report to the accumulator
+		em := errorMetric{}
+		_ = json.Unmarshal([]byte(se.Data), &em)
+		if s.acc != nil {
+			s.wg.Add(1)
+			s.acc.AddError(fmt.Errorf("Received error event for: %v", em.Instance))
+			s.wg.Done()
+		}
+
+	case "keepalive":
+		// Keepalive events are ignored for the moment
+		s.Log.Debug("Received keepalive event")
+
+	default:
+		// Received a yet unsupported event type
+		s.Log.Debug("Received unsupported event")
+	}
+}
+
+// Init is for setup, and validating config.
+func (s *ctrlXDataLayerSSE) Init() error {
+	if s.Timeout == 0 {
+		s.Timeout = defaultTimeout
+	}
+	if time.Duration(s.Timeout) < time.Duration(defaultTimeout) {
+		s.Log.Errorf("Setting timeout must be at least %v seconds", time.Duration(defaultTimeout).Seconds())
+		return errors.New("timeout configuration error")
+	}
```

Comment:
Why do you need this? A timeout of zero means "wait forever", isn't this a valid value? Furthermore, you do not need to log the error manually...
```suggestion
	if time.Duration(s.Timeout) < time.Duration(defaultTimeout) {
		return fmt.Errorf("timeout must be at least %v", defaultTimeout)
	}
```

Reply:
Done